### PR TITLE
core: Add C++ classes for read-only memory-mapped files and UNIX file descriptors; Remove Zstandard's dependency on Boost.

### PR DESCRIPTION
**Pull request by LinZhihao-723**
# Description

## Motivation:
As we plan to publish an npm package for the JavaScript clp-ffi library, we need to remove standard Decompressor's boost dependency to build the necessary components properly. To achieve this milestone, we should replace boost's memory-mapped file with our own implementation. This change is first introduced in PR #387 

## Implementation
This PR introduces two classes necessary to wrap `mmap` system call:
- `FileDescriptor`: This is a wrapper class to a raw file descriptor. With RAII, it is easier to handle failures/exceptions where a raw file descriptor is required. To handle to potential failure in the destructor, we provide a customizable callback if users don't want to swallow the error.
- `MemoryMappedFileView`: This is a wrapper class to `mmap` system call. It maintains the mapped memory region using RAII, and provides methods to access an immutable view of the file. The memory region is unmapped in the destructor.

With the above two classes, we can replace boost's mmap support with our implementation in standard Decompressor.

# Validation performed

1. Ensure the code can be successfully built for all the relevant build targets.
2. Ensure all existing unit tests are passed. The current streaming compression unit tests will call the above implementation to decompress zstd format files.
3. Add new unit tests to test memory-mapped file functionalities separately. It also tests the special case where the mapped file is empty.

### CODE REVIEW

File: components/core/src/clp/FileDescriptor.hpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        std::string m_msg;
+    };
+
+    /**
+     * A C++ wrapper for Unix oflag that describes the operation mode.
+     */
+    // NOLINTNEXTLINE(performance-enum-size)
+    enum class Mode : int {
+        ReadOnly = O_RDONLY,
+        CreateForWrite = O_WRONLY | O_CREAT | O_TRUNC,
```
Comment (junhaoliao) [inline components/core/src/clp/FileDescriptor.hpp:52]: Do we ever need to support append in the future?

File: components/core/tests/test-MemoryMappedFile.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    clp::FileReader file_reader;
+    file_reader.open(test_input_path.string());
+    auto const expected{read_content(file_reader)};
+    file_reader.close();
+
+    clp::MemoryMappedFileView const mmap_file{test_input_path.string()};
+    auto const view{mmap_file.get_view()};
+    std::vector<char> actual;
+    actual.insert(actual.cend(), view.begin(), view.end());
+    REQUIRE((expected == actual));
```
Comment (junhaoliao) [inline components/core/tests/test-MemoryMappedFile.cpp:55]: Are the double braces intended?
```suggestion
    REQUIRE(expected == actual);
```

**Review by junhaoliao — DISMISSED** [review]
lgtm thanks! 

I believe the code changes are sufficient to get rid of the "boost" header dependencies for Emscripten (WASM) compilation for the Zstd decompressor. Looking at cleaning up the CLP repo, can we now also remove references to boost::iostreams in the CMakeList.txt files? (May we need to replace all memeory-mapped file implementations with ours. )

File: components/core/src/clp/FileDescriptor.hpp (hunk 1 above)
Comment (LinZhihao-723) [reply components/core/src/clp/FileDescriptor.hpp:52]: If we need it in the future, we can add more `Mode`

File: components/core/tests/test-MemoryMappedFile.cpp (hunk 2 above)
Comment (LinZhihao-723) [reply components/core/tests/test-MemoryMappedFile.cpp:55]: This is required to silent `clang-tidy` warnings

**Comment by LinZhihao-723** [comment]
> lgtm thanks!
> 
> I believe the code changes are sufficient to get rid of the "boost" header dependencies for Emscripten (WASM) compilation for the Zstd decompressor. Looking at cleaning up the CLP repo, can we now also remove references to boost::iostreams in the CMakeList.txt files? (May we need to replace all memeory-mapped file implementations with ours. )

I agree that we can do more cleaning. However, this is already a non-trivial PR. I think we should delay the rest of the cleanup to the future PRs. Arguably speaking, this PR should only contain two new classes. We should delay the boost deprecation in zstd decompressor to the next PR. What dod u think? @kirkrodrigues

File: components/core/tests/test-MemoryMappedFile.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+#include <Catch2/single_include/catch2/catch.hpp>
+
+#include "../src/clp/FileReader.hpp"
+#include "../src/clp/MemoryMappedFileView.hpp"
+#include "../src/clp/ReaderInterface.hpp"
+
+namespace {
+/**
+ * Reads content from a reader interface till end of file reached.
```
Comment (kirkrodrigues) [inline components/core/tests/test-MemoryMappedFile.cpp:14]:
```suggestion
 * Reads all content from a reader.
```

File: components/core/tests/test-MemoryMappedFile.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+ * @param reader
+ * @return Content read from the given reader.
```
Comment (kirkrodrigues) [inline components/core/tests/test-MemoryMappedFile.cpp:16]:
```suggestion
 * @return The content.
```

File: components/core/tests/test-MemoryMappedFile.cpp (hunk 2 above)
Comment (kirkrodrigues) [inline components/core/tests/test-MemoryMappedFile.cpp:55]: Shall we use `std::equal` instead?

File: components/core/src/clp/MemoryMappedFileView.hpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+#include "ErrorCode.hpp"
+#include "TraceableException.hpp"
+
+namespace clp {
+/**
+ * A class for mapping a read-only file into memory. It maintains the memory buffer created by the
+ * underlying `mmap` system call and provides methods to get a view of the memory buffer.
+ */
+class MemoryMappedFileView {
```
Comment (kirkrodrigues) [inline components/core/src/clp/MemoryMappedFileView.hpp:18]: I think the convention is usually for a view to be a view of some other class. Since this class owns the memory-mapped resource, I feel like calling it a view is a little misleading. I guess you want to call it a view since it only supports read-only files right now. In that case, perhaps it's better to call it a `ReadOnlyMemoryMappedFile`?

File: components/core/src/clp/streaming_compression/zstd/Decompressor.cpp [hunk 6]
```diff
@@ -211,38 +207,11 @@ ErrorCode Decompressor::open(std::string const& compressed_file_path) {
 
     // Create memory mapping for compressed_file_path, use boost read only
     // memory mapped file
```
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_compression/zstd/Decompressor.cpp:209]: Need to update the comment

Comment (kirkrodrigues) [inline components/core/src/clp/streaming_compression/zstd/Decompressor.cpp:209]: Need to update the comment

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
A few things to resolve.

**Comment by kirkrodrigues** [comment]
> However, this is already a non-trivial PR. I think we should delay the rest of the cleanup to the future PRs. Arguably speaking, this PR should only contain two new classes. We should delay the boost deprecation in zstd decompressor to the next PR. What do u think? @kirkrodrigues

Leaving further clean-up to another PR sounds good.

File: components/core/src/clp/MemoryMappedFileView.hpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+    private:
+        std::string m_msg;
+    };
+
+    // Constructors
+    /**
+     * @param path The path of the file to map.
+     */
+    MemoryMappedFileView(std::string_view path);
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/MemoryMappedFileView.hpp:42]: it didn't...

File: components/core/src/clp/FileDescriptor.cpp [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+#include "ErrorCode.hpp"
+#include "type_utils.hpp"
+
+namespace clp {
+FileDescriptor::FileDescriptor(
+        std::string_view path,
+        Mode mode,
+        CloseFailureCallback close_failure_callback
+)
+        : m_fd{open(path.data(), enum_to_underlying_type(mode))},
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/FileDescriptor.cpp:20]: My bad. Will default to `S_IRWXU` if open mode contains `O_CREATE` flag

File: components/core/src/clp/MemoryMappedFileView.hpp (hunk 5 above)
Comment (LinZhihao-723) [reply components/core/src/clp/MemoryMappedFileView.hpp:18]: Yeah make sense. I was about to use this name but I thought you guys may feel it's too verbose

File: components/core/tests/test-MemoryMappedFile.cpp
Comment (kirkrodrigues) [inline components/core/tests/test-MemoryMappedFile.cpp:1]: test-ReadOnlyMemoryMappedFile.cpp?

File: components/core/src/clp/FileDescriptor.cpp [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+#include "type_utils.hpp"
+
+namespace clp {
+FileDescriptor::FileDescriptor(
+        std::string_view path,
+        OpenMode open_mode,
+        CloseFailureCallback close_failure_callback
+)
+        : m_fd{(0 != (enum_to_underlying_type(open_mode) & O_CREAT))
+                       ? open(path.data(), enum_to_underlying_type(open_mode), S_IRWXU)
```
Comment (kirkrodrigues) [inline components/core/src/clp/FileDescriptor.cpp:21]: * `S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH`?
* Also, can we do this in the body of the method instead?
* Can we add a comment for the writable case saying something like:
`// Enable writing for the owner and reading for everyone. Callers can change the created file's permissions as necessary.`

File: components/core/src/clp/FileDescriptor.hpp [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                  m_msg{std::move(msg)} {}
+
+        [[nodiscard]] auto what() const noexcept -> char const* override { return m_msg.c_str(); }
+
+    private:
+        std::string m_msg;
+    };
+
+    /**
+     * A C++ wrapper for Unix oflag that describes the open open_mode.
```
Comment (kirkrodrigues) [inline components/core/src/clp/FileDescriptor.hpp:47]: Overzealous refactoring.

File: components/core/src/clp/FileDescriptor.hpp [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+     */
+    [[nodiscard]] auto get_raw_fd() const -> int { return m_fd; }
+
+    /**
+     * @return The size of the file.
+     */
+    [[nodiscard]] auto get_size() const -> size_t;
+
+    /**
+     * @return The operation open_mode.
```
Comment (kirkrodrigues) [inline components/core/src/clp/FileDescriptor.hpp:82]:
```suggestion
     * @return The open mode.
```

File: components/core/tests/test-MemoryMappedFile.cpp [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+[[nodiscard]] auto get_test_dir() -> std::filesystem::path;
+
+auto read_content(clp::ReaderInterface& reader) -> std::vector<char> {
+    constexpr size_t cBufferSize{4096};
+    std::array<char, cBufferSize> read_buf{};
+    std::vector<char> buf;
+    for (bool has_more_content{true}; has_more_content;) {
+        size_t num_bytes_read{};
+        has_more_content = reader.read(read_buf.data(), cBufferSize, num_bytes_read);
```
Comment (kirkrodrigues) [inline components/core/tests/test-MemoryMappedFile.cpp:29]:
```suggestion
        has_more_content = reader.read(read_buf.data(), read_buf.size(), num_bytes_read);
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/tests/test-MemoryMappedFile.cpp
Comment (LinZhihao-723) [reply components/core/tests/test-MemoryMappedFile.cpp:1]: prefer to name it more generically, in case we want to add writable memory mapped files in the coming future :D

Comment (kirkrodrigues) [reply components/core/tests/test-MemoryMappedFile.cpp:1]: Sure

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

core: Add C++ classes for read-only memory-mapped files and UNIX file descriptors; Remove Zstandard's dependency on boost.

**Comment by LinZhihao-723** [comment]
> For the PR title, how about:
> 
> core: Add C++ classes for read-only memory-mapped files and UNIX file descriptors; Remove Zstandard's dependency on Boost.

lgtm
